src/webapp/api.py

```diff
@@ -0,0 +1,270 @@
+"""
+    This file contains all entry points for the API. The API will use
+    Websockets for most of the communication, although an HTTP route to
+    root has been established, since it makes it easy to check if the
+    application is live.
+"""
+from customvision.classifier import Classifier
+from werkzeug import exceptions as excp
+from flask_socketio import SocketIO, emit, send, join_room
+from flask import request
+from flask import Flask
+from base64 import decodestring, decodebytes
+from PIL import Image
+from io import BytesIO
+from webapp import models
+from utilities.exceptions import UserError
+from utilities import setup
+import logging
+import os
+import json
+import uuid
+import datetime
+
+
+# Initialize app
+app = Flask(__name__)
+logger = False
+if "IS_PRODUCTION" in os.environ:
+    logger = True
+
+socketio = SocketIO(app, cors_allowed_origins="*", logger=logger,)
+app.config.from_object("utilities.setup.Flask_config")
+models.db.init_app(app)
+models.create_tables(app)
+models.seed_labels(app, "./dict_eng_to_nor.csv")
+classifier = Classifier()
+
+
+@socketio.on("connect")
+def connect():
+    print("===== client connected =====")
+
+
+@socketio.on("disconnect")
+def disconnect():
+    """
+        When a player disconnects from a session this function tells the
+        other player in the room who left and deletes all records in the
+        database connected to the session. Old sessions is also deleted.
+    """
+    player_id = request.sid
+    player = models.get_player(player_id)
+    game = models.get_game(player.game_id)
+    data = {
+        "player_disconnected": True
+    }
+    emit("player_disconnected", json.dumps(data), room=game.game_id)
+    models.delete_session_from_game(game.game_id)
+    print("=== client disconnected ===")
+
+
+@socketio.on("message")
+def handle_message(message):
+    print("client: " + str(message))
+
+
+@socketio.on("filetest")
+def handle_filetest(json_data, image):
+    print(json_data)
+    with open("harambe.png", "wb") as f:
+        f.write(image)
+
+
+@socketio.on("joinGame")
+def handle_joinGame(json_data):
+    """
+        Check if player2 is none in mulitplayer table.
+        * If check is false create new mulitplayer game.
+        * If check is true insert player where player2 is none and start
+          the game.
+    """
+    player_id = request.sid
+    # Players join their own room as well
+    join_room(player_id)
+    game_id = models.check_player2_in_mulitplayer(player_id)
+
+    if game_id is not None:
+        # Update mulitplayer table by inserting player_id for player_2 and
+        # change state of palyer_1 in PIG to "Ready"
+        models.update_mulitplayer(player_id, game_id)
+        models.insert_into_players(player_id, game_id, "Ready")
```

Comment:
player_1 does never get the state "ready" - not critical but a bit messy maybe.

Reply:
gidder ikke fikse dette nå